Support multi-dimensional Scatter Assign and Add

Max-K-cadence opened the pull request:
Summary:

This patch generalizes ScatterAssign node. It makes it capable of setting (N-K)
dimensional values by K-dimensional indices to N-dimensional destination.
Without this patch, only K=1 is supported.

It also adds "Cumulative" flag to ScatterAssign instruction. If it is set to
True, then the instruction will add values from Updates to Data by specified
indices instead of copying.

ScatterAssign node is renamed to ScatterData (with optionally set Cumulative
flag) to better reflect its new functionality after changes.

Added support for Interpreter and CPU backends only. OpenCL support can be
added separately by request.

This patch is a preparation step to support of ScatterNd node.

Based on discussion #3153

Test Plan:
Added unit tests. NFCI for existing nodes.

Please see a detailed explanation of how to fill out the fields in the relevant sections in PULL_REQUEST.md.

Max-K-cadence commented:
Looks like some existing models need replacement of ScatterAssign node to ScatterData. What is the location of this /var/lib/jenkins/workspace/downloaded_models/en2gr?

jfix71 commented:
>Looks like some existing models need replacement of ScatterAssign node to ScatterData. What is the location of this /var/lib/jenkins/workspace/downloaded_models/en2gr?

@Max-K-cadence You can find the model in our normal model download script, it's called `en2gr` inside `download_datasets_and_models.py`

Max-K-cadence commented:
I still have no idea what's going on in this script. :D It has nothing but reference to http://fb-glow-assets.s3.amazonaws.com/models/en2gr/ to which I have no access. What should I do to fix this model?

jfix71 commented:
>I still have no idea what's going on in this script. :D It has nothing but reference to http://fb-glow-assets.s3.amazonaws.com/models/en2gr/ to which I have no access. What should I do to fix this model? Is it a part of some different repository?

Not sure what you mean by you don't have access, it should be public. What happens when you run the following?
```
python utils/download_datasets_and_models.py --caffe2-models en2gr
```
It should download all of the files you need for this model to a `en2gr/` directory, and then you should be able to run it via [the instructions here](https://github.com/pytorch/glow/blob/master/docs/Testing.md#text-translation).

Max-K-cadence commented:
My bad, I didn't realize that not the model needed amendment, but the code in Caffe2 loader. Sorry for bothering. :)

jfix71 commented on lib/Graph/Grad.cpp:
```diff
@@ -319,9 +319,9 @@ Function *glow::differentiate(Function *F, const TrainingConfig &conf,
-                                       RG ? RG->getResult() : OutputG);
+      auto *SA = new ScatterDataNode(
+          "scatter.assign.grad", SN->getResult(), RI->getResult(),
+          RG ? RG->getResult() : OutputG, /*Add*/ false);
```
tiny nit: In most other places you're calling it `/*Cumulative*/`. Don't care either way, just prefer to be consistent.

Max-K-cadence replied:
Thanks for pointing out! Yeah, I renamed this during development, and it seems that vestige of old name is still in code. I'll review it carefully.

jfix71 commented on lib/Backends/Interpreter/InterpreterNodes.cpp:
```diff
@@ -1361,25 +1361,140 @@ void BoundInterpreterFunction::fwdGatherRangesInst(
-    size_t destDataIdx = indicesT->getHandle<int64_t>().raw(i);
+  for (size_t i = 0, end = indicesT->dims()[0]; i < end; i++) {
+    size_t destDataIdx = 0;
+    if (indicesT->dims().size() == 1) {
```
Can you add some comments in here about what's going on?

Max-K-cadence replied:
Hm, I think we can simplify it by always turning indices to 2-dimensional.

Max-K-cadence commented:
Thanks Jordan! I'll address the comments and fix the build (now I need to add this node to ONNX writer too).

Max-K-cadence commented:
- Addressed nits;
- Indices are now only 2D for uniformity reasons;
- Added stub to ONNX Exporter. ScatterData has no direct analog in ONNX, but once this one is checked in, I plan to come up with ScatterNd operation that corresponds to it.

jfix71 commented on lib/Backends/CPU/libjit/libjit.cpp:
```diff
@@ -425,16 +425,58 @@ static void libjit_gatherranges(T *output, U *lengths, const T *data,
+                                   size_t sliceSize) {
   for (size_t i = 0; i < numIndices; i++) {
-    size_t destDataIdx = indices[i];
+    size_t destDataIdx = indices[i * indexSize];
```
This seems like it's still generalized past 2D right? Maybe simplify it given we only support 2D?

Max-K-cadence replied:
This code was initially was done for 2D, and it also worked for 1D as `indexSize` as computed as `indices->size() / numIndices`, and for 1D case it was 1. Now I made change in LLVM IR gen to clearly take these values from indices->dims()[0], indices->dims()[1] to make this logic more transparent. But I don't see how we can simplify code here. What's your idea?

jfix71 replied:
Nevermind I think it's good.

jfix71 commented on lib/Backends/Interpreter/InterpreterNodes.cpp:
```diff
@@ -1473,25 +1473,136 @@ void BoundInterpreterFunction::fwdGatherRangesInst(
+    if (indicesT->dims().size() == 1) {
+      destDataIdx = IH.raw(i);
+    } else {
+      assert(indicesT->dims().size() == 2 &&
```
nit: Can move this outside of the loop.

jfix71 commented on lib/Backends/Interpreter/InterpreterNodes.cpp:
```diff
@@ -1473,25 +1473,136 @@ void BoundInterpreterFunction::fwdGatherRangesInst(
+        destDataIdx += IH.at({i, j});
+      }
+    }
+    auto D = dataT->getHandle<ElemTy>(), S = slicesT->getHandle<ElemTy>();
```
nit: Can move this outside of the loop.

Max-K-cadence commented:
Fixed assert for OpenCL.

Max-K-cadence commented:
I'm not sure if it's a real failure or not. I don't think I did something that could cause this bug after fixing OpenMP assert, but I've seen it twice already. :(

jfix71 commented on lib/Graph/Nodes.cpp:
```diff
@@ -1290,32 +1290,30 @@ bool GatherRangesNode::verify() const {
 }
 
-bool ScatterAssignNode::verify() const {
+bool ScatterDataNode::verify() const {
```
Can you add a check here that data and slices should have the same element kind?

jfix71 commented on lib/Backends/Interpreter/InterpreterNodes.cpp:
```diff
@@ -1473,25 +1473,128 @@ void BoundInterpreterFunction::fwdGatherRangesInst(
+  assert(indicesT->dims().size() == 2 &&
+         "Index should be stored in 2D tensor!");
+  const size_t dataSliceSize =
+      slicesT->size() / slicesT->dims()[0] * dataT->getType().getElementSize();
```
tiny nit: Even though they should be the same, I'd use `slicesT` element size instead since we're using its dims/size now just so it's not so confusing when reading this code.

jfix71 commented on tests/unittests/OperatorTest.cpp:
```diff
@@ -2494,6 +2494,238 @@ TEST_P(OperatorTest, ScatterAssignQuantized) {
+
+  bindings_.allocate(data)->getHandle() = {1, 2, 3, 4, 5, 6, 7, 8, 9, 10};
+  bindings_.allocate(indices)->getHandle<int64_t>() = {1, 3};
+  bindings_.allocate(slices)->getHandle() = {-3, -4, -7, -8};
```
Can you make `-8` positive so we verify that the kernels correctly handle maxing out the quantized range?

Max-K-cadence replied:
Hm, looks like they don't. Good catch! I'll fix this.

Max-K-cadence commented:
Thanks for looking through it so thoroughly. :)
